Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean

```diff
@@ -0,0 +1,519 @@
+    at h
+  ext i
+  fin_cases i <;> assumption
+#align upper_half_plane.linear_ne_zero UpperHalfPlane.linear_ne_zero
+
+theorem denom_ne_zero (g : GL(2, ℝ)⁺) (z : ℍ) : denom g z ≠ 0 := by
+  intro H
+  have DET := (mem_glpos _).1 g.prop
+  have hz := z.prop
+  simp only [GeneralLinearGroup.det_apply_val] at DET
+  have H1 : (↑ₘg 1 0 : ℝ) = 0 ∨ z.im = 0 := by simpa [num, denom] using congr_arg Complex.im H
+  cases' H1 with H1
+  · simp only [H1, Complex.ofReal_zero, denom, MulZeroClass.zero_mul, zero_add,
+      Complex.ofReal_eq_zero] at H
+    rw [Matrix.det_fin_two (↑ₘg : Matrix (Fin 2) (Fin 2) ℝ)] at DET
+    simp only [H, H1, MulZeroClass.mul_zero, sub_zero, lt_self_iff_false] at DET
+  · change z.im > 0 at hz
+    linarith
+#align upper_half_plane.denom_ne_zero UpperHalfPlane.denom_ne_zero
+
+theorem normSq_denom_pos (g : GL(2, ℝ)⁺) (z : ℍ) : 0 < Complex.normSq (denom g z) :=
+  Complex.normSq_pos.mpr (denom_ne_zero g z)
+#align upper_half_plane.norm_sq_denom_pos UpperHalfPlane.normSq_denom_pos
+
+theorem normSq_denom_ne_zero (g : GL(2, ℝ)⁺) (z : ℍ) : Complex.normSq (denom g z) ≠ 0 :=
+  ne_of_gt (normSq_denom_pos g z)
+#align upper_half_plane.norm_sq_denom_ne_zero UpperHalfPlane.normSq_denom_ne_zero
+
+/-- Fractional linear transformation, also known as the Moebius transformation -/
+def smulAux' (g : GL(2, ℝ)⁺) (z : ℍ) : ℂ :=
+  num g z / denom g z
+#align upper_half_plane.smul_aux' UpperHalfPlane.smulAux'
+
+theorem smulAux'_im (g : GL(2, ℝ)⁺) (z : ℍ) :
+    (smulAux' g z).im = det ↑ₘg * z.im / Complex.normSq (denom g z) := by
+  rw [smulAux', Complex.div_im]
+  field_simp [smulAux', num, denom]
+  -- porting note: the local notation still didn't work here
+  rw [Matrix.det_fin_two ((g : GL (Fin 2) ℝ) : Matrix (Fin 2) (Fin 2) ℝ)]
+  ring
+#align upper_half_plane.smul_aux'_im UpperHalfPlane.smulAux'_im
+
+/-- Fractional linear transformation, also known as the Moebius transformation -/
+def smulAux (g : GL(2, ℝ)⁺) (z : ℍ) : ℍ :=
+  ⟨smulAux' g z, by
+    rw [smulAux'_im]
+    convert mul_pos ((mem_glpos _).1 g.prop)
+        (div_pos z.im_pos (Complex.normSq_pos.mpr (denom_ne_zero g z))) using 1
+    simp only [GeneralLinearGroup.det_apply_val]
+    ring⟩
+#align upper_half_plane.smul_aux UpperHalfPlane.smulAux
+
+theorem denom_cocycle (x y : GL(2, ℝ)⁺) (z : ℍ) :
+    denom (x * y) z = denom x (smulAux y z) * denom y z := by
+  change _ = (_ * (_ / _) + _) * _
+  field_simp [denom_ne_zero]
+  simp only [Matrix.mul, dotProduct, Fin.sum_univ_succ, denom, num, Subgroup.coe_mul,
+    GeneralLinearGroup.coe_mul, Fintype.univ_ofSubsingleton, Fin.mk_zero, Finset.sum_singleton,
+    Fin.succ_zero_eq_one, Complex.ofReal_add, Complex.ofReal_mul]
+  ring
+#align upper_half_plane.denom_cocycle UpperHalfPlane.denom_cocycle
+
+theorem mul_smul' (x y : GL(2, ℝ)⁺) (z : ℍ) : smulAux (x * y) z = smulAux x (smulAux y z) := by
+  ext1
+  -- Porting note: was `change _ / _ = (_ * (_ / _) + _) * _`
```

Comment:
```suggestion
```
It still is that

Reply:
No, there's a `/` as final operator now, instead of `*`.